lib/input/blob_storage.go

```diff
@@ -0,0 +1,449 @@
+	var timeout time.Duration
+	var err error
+	if tout := conf.Timeout; len(tout) > 0 {
+		if timeout, err = time.ParseDuration(tout); err != nil {
+			return nil, fmt.Errorf("failed to parse timeout period string: %v", err)
+		}
+	}
+
+	if len(conf.StorageAccount) == 0 && len(conf.StorageConnectionString) == 0 {
+		return nil, errors.New("invalid azure storage account credentials")
+	}
+
+	var client storage.Client
+	if len(conf.StorageConnectionString) > 0 {
+		if strings.Contains(conf.StorageConnectionString, "UseDevelopmentStorage=true;") {
+			client, err = storage.NewEmulatorClient()
+		} else {
+			client, err = storage.NewClientFromConnectionString(conf.StorageConnectionString)
+		}
+	} else {
+		client, err = storage.NewBasicClient(conf.StorageAccount, conf.StorageAccessKey)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("invalid azure storage account credentials: %v", err)
+	}
+
+	var objectScannerCtor codec.ReaderConstructor
+	if objectScannerCtor, err = codec.GetReader(conf.Codec, codec.NewReaderConfig()); err != nil {
+		return nil, fmt.Errorf("invalid azure storage codec: %v", err)
+	}
+
+	blobService := client.GetBlobService()
+	a := &azureBlobStorage{
+		conf:              conf,
+		objectScannerCtor: objectScannerCtor,
+		log:               log,
+		stats:             stats,
+		container:         blobService.GetContainerReference(conf.Container),
+		timeout:           timeout,
+	}
+
+	return a, nil
+}
+
+// ConnectWithContext attempts to establish a connection to the target Azure
+// Blob Storage container.
+func (a *azureBlobStorage) ConnectWithContext(ctx context.Context) error {
+	var err error
+	a.keyReader, err = newTargetReader(ctx, a.conf, a.log, a.container, a.timeout)
+	return err
+}
+
+func (a *azureBlobStorage) getObjectTarget(ctx context.Context) (*azurePendingObject, error) {
+	if a.object != nil {
+		return a.object, nil
+	}
+
+	target, err := a.keyReader.Pop(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	blobReference := a.container.GetBlobReference(target.key)
+	exists, err := blobReference.Exists()
+	if err != nil {
+		target.ackFn(ctx, err)
+		return nil, fmt.Errorf("failed to get blob reference: %v", err)
+	}
+
+	if !exists {
+		target.ackFn(ctx, err)
+		return nil, errors.New("blob does not exist")
+	}
+
+	obj, err := blobReference.Get(&storage.GetBlobOptions{Timeout: uint(a.timeout.Seconds())})
+	if err != nil {
+		target.ackFn(ctx, err)
+		return nil, err
+	}
+
+	object := &azurePendingObject{
+		target: target,
+		obj:    blobReference,
+	}
+	if object.scanner, err = a.objectScannerCtor(target.key, obj, target.ackFn); err != nil {
+		target.ackFn(ctx, err)
+		return nil, err
+	}
+
+	a.object = object
+	return object, nil
+}
+
+func blobStorageMsgFromPart(p *azurePendingObject, part types.Part) types.Message {
+	msg := message.New(nil)
+	msg.Append(part)
+
+	meta := msg.Get(0).Metadata()
+
+	meta.Set("blob_storage_key", p.target.key)
```

Comment:
Would adding `meta.Set("blob_storage_filename", path.Base(p.target.key))` be of any use?

Reply:
Nah, it's pretty easy to rename metadata fields with bloblang if anyone needs.